### DISCUSSION TITLE
[4.x] Add Trailing Slash “/” to Joomla URLs

### DISCUSSION THREAD

**Discussion by ptest8794** [joomla/joomla-cms#39213] [General] [answered]
to add trailing slashes in the URL you need to edit the file

### public_html/libraries/src/Router/SiteRouter.php

 public function buildRewrite(&$router, &$uri)
    {
        // Get the path data
        $route = $uri->getPath();

        // Transform the route
        if ($route === 'index.php') {
            $route = '';
        } else {
            $route = str_replace('index.php/', '', $route);
        }

        $uri->setPath($route);
    }

add the string. '/' in the following line of code

### $uri->setPath($route . '/');

I was wondering if it is possible to add as a new feature to be activated from the admin area

**Comment by brianteeman** [comment]
Why would you want to do this?

**Comment by ptest8794** [comment]
To add trailing slash without any plugin or htaccess code.
Do you think is possible add this new feature?

**Comment by richard67** [comment, marked as answer]
Trailing slashes should - if ever - not be added to ANY URL but only to those of directory stile, e.g. "www.some-domain.com/en/" or "www.some-domain.com/de/", but not "www.some-domain.com/index.php/" or "www.some-domain.com/somecategory/somealias.html/". As it is possible to switch on suffix (e.g. ".html") in the configuration, your proposed solution is simply wrong and would break such URLs. Your solution may for you in your particular use case but not for everybody.

**Comment by ptest8794** [comment]
In Fact I mean this solution for cases like www.domain.com/en/category/ or www.domain.com/en/category/article/

I was wondering if it is possbile add this feature with a switch in the web admin area and if the suffix was enabled the trailing slashes button has to set to no add or disabled.

**Comment by richard67** [comment]
Well the trailing slashes are a relict from old times. Meanwhile Google doesn't recommend or require them anymore. They do not do any harm, but they also serve no benefit.

**Comment by joomdonation** [comment]
Guess this issue could be closed. @richard67 Could you please make final decision here?

**Comment by richard67** [comment]
@joomdonation I don’t like to make decisions. Maybe move it to discussions?

**Comment by joomdonation** [comment]
OK. Thanks. I'm moving this to discussions so that it could be discussed further if needed.

**Comment by jokorntheuer** [comment]
As an SEO-Senior i can tell you that this issue DOES matter. Why?
First of all - the best solution would be to make this a GLOBAL OPTION in joomla -"Trailing slash - YES / NO" - for EVERY URL, no matter if it is a category or article.

Google recommends to use EITHER no trailing slash oWITH trailing slash and RECOMMENDS to be consistent here.
And there comes the **problem** - **Joomla 4 URLS can be rendered WITH Trailing slashes and without trailing slashes - this can produce cuplicate content and is NOT CONSISTENT.**
Source: https://developers.google.com/search/blog/2010/04/to-slash-or-not-to-slash?hl=de

Also this would help me (current project in fact) to migrate a Wordpress-Site to Joomla. Because wordpress uses Trailing Slashes "/" on every url and joomla does not. So i would have to make a 301-redirect for every single site (with relevant rankings) to the new joomla URL with no trailing slash. If i dont do this - the old Wordpress URLS would also be working on the joomla ist (if URL structure is the same) and produce URLS with Trailing slash (from old wordpress urls who are still in the google index) and without trailing slash (internal joomla linking).

**joomla 4 needs the following SEO-Settings:**
- global option: Trailing slash: YES OR NO
- correct automatic redirects of URLS with trailing slashes TO URLS without trailing slashes and VICE VERSA depending on the global setting.

in my opinion this is a "design-flaw" that should have been adressed already in joomla 3 and i am really surprised why this isnt yet fixed in joomla 4 because it makes the lifes and work of seo-agencies much more difficult and is not a [+] point when it comes to the choice "wordpress OR Joomla".

greetings,
Joe

**Reply by brianteeman** [reply]
This is a typical comment from an SEO expert

**Reply by brianteeman** [reply]
To expand on that comment. Just because wp does something it doesnt mean it is correct and the only way. The fact that wp puts a slash at the end of very url doesnt mean it is correct. The reality is that wp is an outlier on this and the majority of sites do not do this. But if you insist that it is better for selling snake oil then you can easily write one htaccess rule for the entire site

**Reply by webgras** [reply]
@brianteeman it is not that easy, because with the htaccess rule you produce internal redirects, also not good for SEO, if you want a clean link structure.
Everytime I have SEO-partners in a project, we stumble on that issue. And yes, it is a wp issue, but yes, wp is the majority of cms sites nowadays.
I would really vote for that option - to turn it on, if you want it.

**Reply by jokorntheuer** [reply]
@brianteeman you re ignoring that Joomla (current version) doesnt do it right either. urls need to be consistant and should not be possible to be rendered WITH and WITHOUT trailing slashes. why not make this option possible in joomla and let users decide?
and speaking of google - have you checked my link? --> https://developers.google.com/search/blog/2010/04/to-slash-or-not-to-slash?hl=de

there it is clearly described WHY it is not GOOD to have URLS that resolve with and without trailing slashes.

and besides being an SEO expert - i am a big joomla fan too and constantly defending joomla as CMS when it comes to new project in our web/seo agency. but these flaws need to be taken serious.

**Reply by jokorntheuer** [reply]
and to expand on that comment - THE majority of sites generats SEF-URLS that are not rendered with "/" and without trailing-slash. Drupal for example also automatically redirects trailing-slash urls to non-trailing-slash urls.

**Reply by bembelimen** [reply]
To make it easy: if it is a absolutely needed feature for all SEOs, it shouldn't be a problem to throw some money in a pot, hire a programmer who implements this in the core.

**Reply by jokorntheuer** [reply]
@bembelimen it is an absolutely needed feature for all CMS who want to be taken serious and be considered to be chosen for web-projects where seo-rankings matter.

**Reply by bembelimen** [reply]
Like I said, if the feature is important for you, then you have all the options to get it implemented by e.g. hiring a developer to contribute to the core.

**Reply by brianteeman** [reply]
I strongly suggest that you read ALL of the linked article. https://www.danielmorell.com/guides/htaccess-seo/redirects/https-www-and-trailing-slash and not rely on an article from 2010

**Reply by webgras** [reply]
I read the whole article in detail. Ok, there is a solution for to have just one override with the proposed htaccess-lines and Google says it is okay, to have up to 5 internal redirects. 
Still, why don't we add an option to add trailing slashes?
@brianteeman why are you against an option like this?

**Reply by brianteeman** [reply]
> @brianteeman why are you against an option like this?

did I ever say I was?

but now that you ask. I am against adding pointless options to Joomla. They just confuse users and add extra code for people to maintain. As I have shown (and you agree) an option is not needed as you can simply add a simple directive to your own htaccess file. But that is just my personal opinion which in the absence of any code is just hypothetical

**Reply by jokorntheuer** [reply]
@brianteeman - i ve read the article too. Many things are correct but you re missing the whole point tbh.
Also, i HAVE NOT SAID that Wordpress does the SEF-Routing BETTER than Joomla or IS BEST at all. What i have said is, that Wordpress uses (at least) consistant URL-Routing AND it would make it easier (SEO-wise) when u migrate a WP-Site to Joomla if such an option would exist. Wordpress SEF-Urls have only 1 Version - URLS with trailing slash. Point. In joomla - we have both or even more. And that is the problem. If it is possible to make Joomla add "/" at the end for "directories" and no trailing-slash when it is an article - please do so! But then this solution should be unique and consistant for the whole site and duplicate urls (with and without trailing slash) shouldnt be possible.

And what you forget is that this problem currently also negatively impact the way how canonical links are handled in joomla.
Right now the canonical-url is EXACTLY the same as the SEF-URL. Example:
www.yourjoomlasite.com/content - canonical: www.yourjoomlasite.com/content
www.yourjoomlasite.com/content/ - canonical: www.yourjoomlasite.com/content/
www.yourjoomlasite.com/index.php/content - canonical: www.yourjoomlasite.com/index.php/content
www.yourjoomlasite.com/index.php/content/ - canonical: www.yourjoomlasite.com/index.php/content/

and this is WRONG. The canonical link needs to BE UNIQUE to tell google which URL should be indexed.
And because joomla sef-urls can be accessed via multiple SEF-variations, we have X canonical-links (status quo).

CORRECT solution would be:
ALL 4 url-examples (i meantioned before) should produce ONLY ONE canonical URL. --> the one url i want google to index.

BEST solution:
ONLY 1 URL (not 4 variations) produces ONE canonical URL.

The correct / best solution is currently not possible with joomla core version 4.2.9.

If there is a magical HTACCESS-Rule that makes this all happen for joomla why hasnt this implemented yet?

From an SEO-perpective Joomla needs to improve. Would you agree with me on this, or no?

**Reply by brianteeman** [reply]
> Like I said, if the feature is important for you, then you have all the options to get it implemented by e.g. hiring a developer to contribute to the core.

**Reply by jokorntheuer** [reply]
So to sum this up:
the BIG problem here is the combination of joomla producing multiple SEF-URLS TOGETHER with MULTIPLE Canonical-URLS.

**Reply by jokorntheuer** [reply]
hey see. i can do quotes too.
> @brianteeman it is an absolutely needed feature for all CMS who want to be taken serious and be considered to be chosen for web-projects where seo-rankings matter.

**Reply by jokorntheuer** [reply]
that`s it for me now. i dunno why i am investing any time here to report an issue with details and how it could improve the value of joomla (compared to other CMS who do it better) just to get quotes like: "Hire a developer".

**Reply by brianteeman** [reply]
/me still looking for a canonical link in joomla

**Reply by jokorntheuer** [reply]
> /me still looking for a canonical link in joomla

no canonical link in joomla core? ok. this is serious. maybe this comes from PWT SEO. nevertheless - "canonical link" is just another "must have feature for CMS in the year 2023" shall i open a new issue?

**Reply by brianteeman** [reply]
so you are saying that the _problem_ with your multiple canonical links is because of an extension to Joomla creating them. In which case you should report it there

**Reply by jokorntheuer** [reply]
oh well, i found another topic on this...already from...2019. @Hackwar pinned the problem in the last comment. the topic creator btw is digging in on the same problem that still exists 4 years later:
https://github.com/joomla/joomla-cms/issues/25448

**Comment by Hackwar** [comment]
As @richard67 wrote, we shouldn't have the slashes at the end and Joomla does not render the URLs with a slash at the end. We are not going to add this as a switch in the configuration somewhere. We already have far to many switches and those create additional issues. If you need a slash at the end, you can write a plugin to do this. I'm going to close this discussion with that.
